src/js/common/schemaform/ObjectField.jsx

```diff
@@ -49,29 +36,13 @@ class ObjectField extends React.Component {
     this.onPropertyChange = this.onPropertyChange.bind(this);
     this.onPropertyBlur = this.onPropertyBlur.bind(this);
     this.isRequired = this.isRequired.bind(this);
-    // This runs a series of steps that order properties and then group them into
-    // expandable groups. If there are no expandable groups, then the end result of this
-    // will be an array of single item arrays
-    this.orderAndFilterProperties = _.flow(
-      properties => orderProperties(properties, _.get('ui:order', this.props.uiSchema)),
-      _.groupBy((item) => {
-        const expandUnderField = _.get([item, 'ui:options', 'expandUnder'], this.props.uiSchema);
-        return expandUnderField || item;
-      }),
-      _.values
-    );
+    this.SchemaField = pureWithDeepEquals(this.props.registry.fields.SchemaField);
+    this.orderedProperties = this.orderAndFilterProperties(props.schema, props.uiSchema);
   }
 
   componentWillReceiveProps(nextProps) {
-    const state = this.getStateFromProps(nextProps);
-    const { formData } = nextProps;
-    if (formData && objectKeysHaveChanged(formData, this.state)) {
-      // We *need* to replace state entirely here has we have received formData
-      // holding different keys (so with some removed).
-      this.state = state;
-      this.forceUpdate();
-    } else {
-      this.setState(state);
+    if (this.props.schema !== nextProps.schema || this.props.uiSchema !== nextProps.uiSchema) {
```

Comment:
We only need to recalculate the list of properties if the schema have changed, so we can store the list on the component instance and only update it when necessary.